operator/cmd/metrics/adapter/provider/custom_metrics.go

```diff
@@ -0,0 +1,106 @@
+package provider
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/metrics/pkg/apis/custom_metrics"
+	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
+)
+
+var ErrMetricNotFound = errors.New("no metric value found")
+
+type CustomMetricValue struct {
+	Value  custom_metrics.MetricValue
+	Labels map[string]string
+}
+
+type CustomMetricsCache struct {
+	mtx     sync.RWMutex
+	metrics map[string]CustomMetricValue
+}
+
+// Update adds a new metricValue for the given metricName to the cache. If an item has already been present for the provided
+// metricName, the previous value will be replaced.
+func (cm *CustomMetricsCache) Update(metricName string, metricValue CustomMetricValue) {
+	cm.mtx.Lock()
+	defer cm.mtx.Unlock()
+	if cm.metrics == nil {
+		cm.metrics = map[string]CustomMetricValue{}
+	}
+
+	cm.metrics[metricName] = metricValue
+}
+
+// Delete will delete the value for the given metricName
+func (cm *CustomMetricsCache) Delete(metricName string) {
+	cm.mtx.Lock()
+	defer cm.mtx.Unlock()
+
+	delete(cm.metrics, metricName)
+}
+
+// List returns a slice of provider.CustomMetricInfo objects containing all the available metrics
+// that are currently present in the cache
+func (cm *CustomMetricsCache) List() []provider.CustomMetricInfo {
+	cm.mtx.RLock()
+	defer cm.mtx.RUnlock()
+	res := []provider.CustomMetricInfo{}
```

Comment:
if I understand it correctly, we know the number of items that will be in the array, can we allocate the array directly for the length and capacity?
```suggestion
	res := make([]provider.CustomMetricInfo, 0, len(cs.metrics))
```

Reply:
yep, here we can do it this way